### PR TITLE
updated source position function.

### PR DESCRIPTION
**Pull request by nkhadka21**

### CODE REVIEW

File: slsim/Sources/source.py [hunk 1]
```diff
@@ -123,3 +123,53 @@ def extended_source_magnitude(self, band):
...
+    def extended_source_position(self, deflector_center, test_area):
+        """Extended source position. If not present from the catalog, it is drawn
+        uniformly within the circle of the test area centered on the deflector position.
+
+        :param deflector_center: center of the deflector.
+         Eg: np.array([center_x_lens, center_y_lens])
+        :return: [x_pos, y_pos]
+        """
+        self.center_lens = deflector_center
+        self.test_area = test_area
```
Comment (sibirrer) [inline slsim/Sources/source.py:136]: no need to make this a class variable

File: slsim/Sources/source.py [hunk 2]
```diff
@@ -123,3 +123,53 @@ def extended_source_magnitude(self, band):
             band_string = "mag_" + band
         source_mag = self.source_dict[band_string]
         return source_mag
+
+    def extended_source_position(self, deflector_center, test_area):
+        """Extended source position. If not present from the catalog, it is drawn
+        uniformly within the circle of the test area centered on the deflector position.
+
+        :param deflector_center: center of the deflector.
```
Comment (sibirrer) [inline slsim/Sources/source.py:131]: doc string for 'test_area' missing. Perhaps also re-name this variable for this function, as conceptually in this function, it has a different purpose.

File: slsim/Sources/source.py [hunk 3]
```diff
@@ -123,3 +123,53 @@ def extended_source_magnitude(self, band):
...
+    def point_source_position(self, deflector_center, test_area):
+        """Point source position. point source could be at the center of the extended
+        source or it can be off from center of the extended source.
+
+        :param deflector_center: center of the deflector.
+         Eg: np.array([center_x_lens, center_y_lens])
+        :return: [x_pos, y_pos]
+        """
+        self.center_lens = deflector_center
+        self.test_area = test_area
```
Comment (sibirrer) [inline slsim/Sources/source.py:159]: same comments as for the other definition related to 'test_area'

File: slsim/lens.py [hunk 4]
```diff
@@ -558,7 +577,9 @@ def source_light_model_lenstronomy(self, band=None):
         """
         source_models = {}
         all_source_kwarg_dict = {}
-        center_source = self.source_position
+        center_source = self.source.extended_source_position(
```
Comment (sibirrer) [inline slsim/lens.py:580]: does this work/not crash if we only have a point source without an extended image?

**Review by sibirrer — CHANGES_REQUESTED** [review]
Thank you @nkhadka21. I have a few minor comments and will go through it again in more detail when I find more time

File: slsim/lens.py (hunk 4 above)
Comment (nkhadka21) [reply slsim/lens.py:580]: I moved this inside the if statement.

**Comment by nkhadka21** [comment]
Hi @sibirrer, resolved all the issues.

File: slsim/lensed_system_base.py [hunk 5]
```diff
@@ -46,20 +46,16 @@ def deflector_position(self):
...
         """
         pass
 
     @abstractmethod
-    def image_positions(self):
-        """Returns image positions by solving the lens equation, these are either the
-        centers of the extended source, or the point sources in case of (added) point-
-        like sources, such as quasars or SNe.
+    def point_source_image_positions(self):
+        """Returns point source image positions by solving the lens equation.
```
Comment (sibirrer) [inline slsim/lensed_system_base.py:58]: also here, this function may return the center of the extended source in the absence of a point source

**Review by sibirrer — CHANGES_REQUESTED** [review]
Thank you very much! I have a few more minor requests to have a more specific code and structure.

File: slsim/Sources/source.py [hunk 6]
```diff
@@ -123,3 +123,52 @@ def extended_source_magnitude(self, band):
...
+        :param draw_area: The area of the test region from which we randomly draw a
+         source position. Eg: 4*pi.
+        :return: [x_pos, y_pos]
+        """
+
+        extended_source_center = self.extended_source_position(center_lens, draw_area)
+
+        if "ra_off" in self.source_dict.colnames:
+            center_x_point_source = (
+                extended_source_center[0] + self.source_dict["ra_off"][0]
```
[the remark this answers is not in the document]
Comment (nkhadka21) [reply slsim/Sources/source.py:165]: You are right. corrected!

File: slsim/lens.py [hunk 7]
```diff
@@ -180,13 +183,27 @@ def validity_test(
         # Criteria 3: The distance between the lens center and the source position
         # must be less than or equal to the angular Einstein radius
         # of the lensing configuration (times sqrt(2)).
-        center_lens, center_source = self.deflector_position, self.source_position
+        if self._source_type == "extended":
```
[the remark this answers is not in the document]
Comment (nkhadka21) [reply slsim/lens.py:186]: You are correct! we do not need this.

**Comment by nkhadka21** [comment]
Hi @sibirrer , resolved all the comments!

**Review by sibirrer — APPROVED** [review]
Thank you @nkhadka21

**Pull request merged**